Add confirmation dialog when performing dangerous actions in tournament client

ILW8 opened the pull request:
Many "dangerous" actions in the tournament client do not require confirmation at all, and the buttons often are located near other oft-used buttons in the various editor screens. 

For example: 
- `Reset teams` button in the bracket editor screen's context menu
- `Clear all` button in the control panel of both Rounds editor and Team editor screens
- `Delete Round` button in Rounds editor screen
- `Delete Team` button in Team editor screen

This PR implements confirmation dialog for these dangerous actions and is mostly done by "copying" some of the dialog overlay code from OsuGame into TournamentGame.


This also adds tests to test the new dialogs. I couldn't find an existing example of a similar functionality nor test in the tournament client tests so I kinda winged it, hoping that it still meets the project's code quality standards.


- Feature
  - [x] `Reset teams` button in the bracket editor screen's context menu
  - [x] `Clear all` button in the control panel of both Rounds editor and Team editor screens
  - [x] `Delete Round` button in Rounds editor screen
  - [x] `Delete Team` button in Team editor screen
  - [x] add dialog overlay to editor screen test scenes

- Tests:
  - [x] test `Reset teams` button in the bracket editor screen's context menu

ILW8 committed:
add dialogoverlay to tournament client, confirmation dialog for resetting bracket teams

ILW8 committed:
refactor LadderResetTeamsDialog.cs

ILW8 committed:
add ladder editor reset teams confirmation dialog test

ILW8 committed:
reorder LadderEditorScreen tests

ILW8 committed:
move LadderResetTeamsDialog.cs into editor screens components

ILW8 committed:
add confirmation dialog for Clear All button in TournamentEditorScreen

ILW8 committed:
remove unused code

ILW8 committed:
fix ResetBracketTeams tests

also fixes formatting and var naming issues

bdach commented:
When it comes to testing tournament client there's probably no need to overdo it and test it zealously. If the changes you make to add the confirmation dialogs are obvious (which they should be), we should be fine without testing.

bdach commented on osu.Game.Tournament.Tests/Screens/TestSceneLadderEditorScreen.cs:
```diff
@@ -1,22 +1,99 @@
 // See the LICENCE file in the repository root for full licence text.
 
-using osu.Framework.Allocation;
+#nullable disable
```
Please don't use `#nullable disable` in new code unless there's a very good reason to do so.

bdach commented on osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public partial class LadderResetTeamsDialog : DangerousActionDialog
+    {
+        public LadderResetTeamsDialog(Container<DrawableTournamentMatch> matchesContainer)
```
Please don't pass containers to these dialogs like this, pass the `Action` to execute instead.

Hell I'm not sure you even need to make all these classes. You could probably use `DangerousActionDialog` on its own and locally set `HeaderText`, `Icon` and `DangerousAction` as you see fit.

ILW8 replied:
`DangerousActionDialog` is declared abstract and its constructor as well as DangerousAction are protected. It looks like `DangerousAction` wasn't intended to be used directly and was meant to be subclassed. 
Of course I could remove the `abstract` modifiers and change both ctor and DangerousAction to public, thoughts?

bdach replied:
I guess I didn't notice _that_, somehow.

Probably keep the classes. Note about passing `Action` still applies, though.

bdach commented on osu.Game.Tournament/Screens/Editors/Components/LadderResetTeamsDialog.cs:
```diff
@@ -0,0 +1,24 @@
+    {
+        public LadderResetTeamsDialog(Container<DrawableTournamentMatch> matchesContainer)
+        {
+            HeaderText = @"Confirm reset teams?";
```
```suggestion
            HeaderText = @"Reset teams?";
```

I can't say I like this phrasing with "confirm" jammed in there at the start. Reads wrong.

bdach commented on osu.Game.Tournament/Screens/Editors/Components/TournamentClearAllDialog.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public TournamentClearAllDialog(IList storage)
+        {
+            HeaderText = @"Confirm clear all?";
```
```suggestion
            HeaderText = @"Clear all?";
```

Mind you this text is still wonderfully unspecific but making it specific is going to be somewhat painful due to surrounding structure of `TournamentEditorScreen` so probably fine to stop at the generic option...

bdach commented on osu.Game.Tournament/Screens/Editors/LadderEditorScreen.cs:
```diff
@@ -30,6 +33,10 @@ public partial class LadderEditorScreen : LadderScreen, IHasContextMenu
 
         private WarningBox rightClickMessage;
 
+        [Resolved]
+        [CanBeNull]
```
This - unfortunately - doesn't quite work like you'd like it too. Yes the `[CanBeNull]` technically means that resharper will null-check the property, _however_ `[Resolved]` without `CanBeNull = true` set inside means that DI will crash if it fails to resolve the dialog overlay. So in practice, `dialogOverlay` will either not be null, or the game will crash.

The proper definition is

```csharp
            [Resolved(canBeNull: true)]
            [CanBeNull]
            private IDialogOverlay dialogOverlay { get; set; }
```

With NRT this gets a lot less boilerplatey, but we can't go there yet in tournament client.

bdach commented on osu.Game.Tournament/Screens/Editors/TournamentEditorScreen.cs:
```diff
@@ -27,6 +30,10 @@ public abstract partial class TournamentEditorScreen<TDrawable, TModel> : Tourna
     {
         protected abstract BindableList<TModel> Storage { get; }
 
+        [Resolved]
+        [CanBeNull]
+        private IDialogOverlay dialogOverlay { get; set; }
```
As above

bdach commented on osu.Game.Tournament/TournamentGame.cs:
```diff
@@ -90,12 +101,14 @@ private void load(FrameworkConfigManager frameworkConfig, GameHost host)
                     loadingSpinner.Hide();
                     loadingSpinner.Expire();
 
+                    topMostOverlayContent.Add(dialogOverlay);
```
This is weird, why is `dialogOverlay` synchronously added on the load complete callback of the _async load_? Can you put the `dialogOverlay` as a child of `topMostOverlayContent` from the start?

ILW8 replied:
This is the part I was kinda winging, I'm not sure how I ended up with the code in this form; IIRC doing what felt intuitive to me did not seem to work. I'll incorporate your other suggested changes first and revisit this soon

ILW8 committed:
remove #nullable disable, pass action instead of container to DangerousActionDialog

ILW8 committed:
fix weird handling of dialogOverlay in tourney client

ILW8 committed:
pass action instead of storage to TournamentClearAllDialog

ILW8 committed:
add confirmation dialog to delete team/round buttons in tourney editor screens

bdach committed:
Merge branch 'master' into tourney-dangerous-action-confirm

bdach committed:
Touch up dialogs a bit

bdach committed:
Remove superfluous `CanBeNull` dependency specs

As stated before, not required in NRT-annotated files if auto-properties
are already marked nullable.

bdach committed:
Remove superfluous container

peppy committed:
Merge branch 'master' into tourney-dangerous-action-confirm

peppy committed:
Cache `IDialogOverlay` in a simpler way
